src/flist/mount.rs

```diff
@@ -0,0 +1,487 @@
+    }
+
+    // returns the mount path out of an flist name simplly joins /<FLISTS_ROOT>/<mountpoint>/<name>
+    // this where this flist instance will be
+    pub fn mountpath<T: AsRef<str>>(&self, name: T) -> Result<PathBuf> {
+        let mountpath = self.mountpoint.join(name.as_ref());
+        if mountpath.parent() != Some(self.mountpoint.as_path()) {
+            bail!("inavlid mount name: {}", name.as_ref());
+        }
+        Ok(mountpath)
+    }
+
+    // returns ro path joined with flist hash
+    // this where we mount the flist for read only
+    fn flist_ro_mount_path<R: AsRef<str>>(&self, hash: R) -> Result<PathBuf> {
+        let mountpath = self.ro.join(hash.as_ref());
+        if mountpath.parent() != Some(self.ro.as_path()) {
+            bail!("invalid mount name")
+        }
+
+        Ok(mountpath)
+    }
+
+    // Checks if the given path is mountpoint or not
+    pub async fn is_mounted<P: AsRef<Path>>(&self, path: P) -> bool {
+        matches!(storage::mountpoint(path.as_ref()).await, Ok(Some(_)))
+    }
+
+    // Checks is the given path is a valid mountpoint means:
+    // it is either doesn't exist or
+    // it is a dir and not a mountpoint for anything
+    pub async fn valid<P: AsRef<Path>>(&self, path: P) -> bool {
+        match fs::metadata(&path).await {
+            Ok(info) => info.is_dir() && !self.is_mounted(&path).await,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => true,
+            Err(err) if err.kind() == io::ErrorKind::ConnectionAborted => {
+                matches!(self.syscalls.umount(path, None), Ok(_))
+            }
+            _ => false,
+        }
+    }
+
+    async fn wait_mountpoint<P: AsRef<Path>>(&self, path: P, seconds: u32) -> Result<()> {
+        let mut duration = seconds;
+        while duration > 0 {
+            time::sleep(time::Duration::from_secs(1)).await;
+            if self.is_mounted(path.as_ref()).await {
+                return Ok(());
+            }
+            duration -= 1;
+        }
+
+        bail!("{}, was not mounted in time", path.as_ref().display())
+    }
+
+    // MountRO mounts an flist in read-only mode. This mount then can be shared between multiple rw mounts
+    // TODO: how to know that this ro mount is no longer used, hence can be unmounted and cleaned up?
+    // this mounts the downloaded flish under <FLISTS_ROOT>/ro/<FLIST_HASH>
+    pub async fn mount_ro<T: AsRef<str>, W: AsRef<str>>(
+        &self,
+        url: T,
+        storage_url: Option<W>,
+    ) -> Result<PathBuf> {
+        // this should return always the flist mountpoint. which is used
+        // as a base for all RW mounts.
+        let flist_path = self.db.get(url).await?;
+
+        let hash = match flist_path.file_name() {
+            Some(hash) => match hash.to_str() {
+                Some(hash) => hash,
+                None => bail!("failed to get flist hash"),
+            },
+            None => bail!("failed to get flist hash"),
+        };
```

Comment:
Instead of doing this, wouldn't it be better if the `self.db.get(url)` returns both the hash and the url separately like
```rust
let (hash, path) = self.db.get(url).await?;
```